### DISCUSSION TITLE
Azure ACR Auto Login - Service Principal

### DISCUSSION THREAD

**Discussion by clementblaise** [fluxcd/flux2#2488] [Q&A] [answered]
Hello,

The [documentation](https://fluxcd.io/docs/guides/image-update/#using-native-azure-acr-auto-login) is pretty clear on how to setup auto login with Azure ACR using managed identity.

I am wondering if it is possible to use the same feature with a Service Principal ? I currently use this mechanism to allow the Kubelet to pull image from ACR.

**Comment by kingdonb** [comment]
Hello @clementblaise I think this is a gap. There are a few mentions of Service Principal in the azure guide at-large, but no examples. There is an example in the SOPS guide which mentions storing Azure Service Principal credentials in Kubernetes secrets:

https://fluxcd.io/docs/guides/mozilla-sops/#azure

I am fairly certain it is supported, and has been tested to work, but I have not tested it. If you can report back on your findings, maybe this documentation will help you get over the finish line; it would be great to contribute back a note in the Azure guide to make it clearer how to use Service Principal directly! (My test environment disabled service principals or I would have contributed this myself.)

**Comment by clementblaise** [comment, marked as answer]
Thanks @kingdonb but SOPS decryption is done by the `kustomize-controller `. My question was regarding the ImageRepository which is manage by the `image-reflection-controller`

**Reply by kingdonb** [reply]
I understand that, however I'm assuming that a service principal is a service principal and that's the reason why it hasn't been documented separately again. I'm suggesting to create a secret in the cluster representing your service principle in the same manner as SOPS docs describes (the service principal information is linked through the first paragraph of the Flux docs that I linked, see: https://github.com/mozilla/sops#encrypting-using-azure-key-vault) and attach it to the image-reflector-controller as described by those docs for kustomize-controller:

> if using a Service Principal stored in a K8s Secret, patch the Pod’s envFrom to reference the `AZURE_TENANT_ID`/`AZURE_CLIENT_ID`/`AZURE_CLIENT_SECRET` fields from your Secret.

I'm saying this is apparently a gap in our docs, so you will need to fill in some blanks... but "in principle" the service principal attachment should be the same no matter which controller needs the credentials bound to it.

**Reply by clementblaise** [reply]
Thanks for the precision. I preferred to go on an alternative option, I have deployed a pull secret and kept all the ImageRepository/ImagePolicy in the same namespace.
